cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -246,6 +278,52 @@ public void run() {
     }
   }
 
+  private TransactionContext getTransactionContextFromContext(WorkflowContext context) {
+    TransactionContext transactionContext = new TransactionContext(txClient);
+    Collection<Dataset> datasets = ((BasicWorkflowContext) context).getDatasets().values();
+    for (Dataset ds : datasets) {
+      transactionContext.addTransactionAware((TransactionAware) ds);
+    }
+    return transactionContext;
+  }
+
+  private void runInTransaction(WorkflowAction action, WorkflowContext workflowContext) {
+    TransactionContext transactionContext = getTransactionContextFromContext(workflowContext);
+    try {
+      transactionContext.start();
+      action.run();
```

Comment:
Also, you can reduce the code duplication by have a method to execute transactionally.

```<V> V executeInTransaction(WorkflowContext workflowContext, Callable<V> callable) throws TransactionFailureException```